intf/intf.go

```diff
@@ -154,3 +168,56 @@ func AddIP(intf string, addr *net.IPNet) error {
 	}
 	return nil
 }
+
+// AwaitARP waits for the IPv4 address addr to be resolved via ARP. It uses the supplied
+// context to determine whether it should continue awaiting the entry. If the ARP entry
+// is resolved, it is returned directly.
+func AwaitARP(ctx context.Context, addr net.IP) (net.HardwareAddr, error) {
+	neighs, err := accessor.ARPList()
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve list of ARP entries, %v", err)
+	}
+
+	for _, n := range neighs {
+		if addr.Equal(n.IP) {
+			return n.MAC, nil
+		}
+	}
+
+	updates := make(chan ARPUpdate, 1)
+	result := make(chan ARPUpdate, 1)
+	done := make(chan struct{})
+
+	go func(updates chan ARPUpdate, done chan struct{}) {
+		for {
+			select {
+			case upd := <-updates:
+				if upd.Type == ARPAdd {
+					if addr.Equal(upd.Neigh.IP) {
+						result <- upd
+					}
+				}
+			case <-done:
+				return
+			}
+		}
+	}(updates, done)
+
+	if err := accessor.ARPSubscribe(updates, done); err != nil {
+		return nil, fmt.Errorf("cannot subscribe to ARP updates, err: %v", err)
+	}
+
+	select {
+	case n := <-result:
+		done <- struct{}{}
+		return n.Neigh.MAC, nil
+	case <-ctx.Done():
```

Comment:
is this case trigger from doing ctx.WithTimeout()? I assume after 30 seconds, an empty struct is put into the ctx.Done() channel?

Reply:
That's correct -- `ctx.Done()` returns a channel that is written to when that channel becomes complete - e.g., the timeout fires, or the cancellation function is called.